diff --git a/Plugins/Wwise/Source/WwiseResourceCooker/Private/Wwise/WwiseResourceCookerImpl.cpp b/Plugins/Wwise/Source/WwiseResourceCooker/Private/Wwise/WwiseResourceCookerImpl.cpp
index 9cee64f..013f8a0 100644
--- a/Plugins/Wwise/Source/WwiseResourceCooker/Private/Wwise/WwiseResourceCookerImpl.cpp
+++ b/Plugins/Wwise/Source/WwiseResourceCooker/Private/Wwise/WwiseResourceCookerImpl.cpp
@@ -922,6 +922,13 @@ void FString FWwiseResourceCookerImpl::GetCookRootPath(const TCHAR* PackageFilename)
 		Path = FPaths::GetPath(Path);
 	}
 
+	while (!Path.IsEmpty() && FPaths::GetCleanFilename(Path) != FApp::GetProjectName())
+	{
+		Path = FPaths::GetPath(Path);
+	}
+
+	Path = Path / TEXT("Content");
+
 	return Path;
 }
 
