diff --git a/Plugins/Wwise/Source/AkAudio/Private/AkWaapiClient.cpp b/Plugins/Wwise/Source/AkAudio/Private/AkWaapiClient.cpp
index 9dfaeac..f4ecb73 100755
--- a/Plugins/Wwise/Source/AkAudio/Private/AkWaapiClient.cpp
+++ b/Plugins/Wwise/Source/AkAudio/Private/AkWaapiClient.cpp
@@ -73,7 +73,7 @@
 
 void FAkWaapiClientConnectionHandler::RegisterAutoConnectChangedCallback()
 {
-#if WITH_EDITOR
+#if AK_SUPPORT_WAAPI
 	FScopeLock Lock(&AkSettingsSection);
 	if (auto AkSettingsPerUser = GetDefault<UAkSettingsPerUser>())
 	{
