diff --git a/Plugins/Wwise/Source/AkAudio/Private/AkAudioDevice.cpp b/Plugins/Wwise/Source/AkAudio/Private/AkAudioDevice.cpp
index d570163eb3..400925ec4d 100644
--- a/Plugins/Wwise/Source/AkAudio/Private/AkAudioDevice.cpp
+++ b/Plugins/Wwise/Source/AkAudio/Private/AkAudioDevice.cpp
@@ -103,7 +103,7 @@ Copyright (c) 2024 Audiokinetic Inc.
 #include <inttypes.h>
 
 //<CSS> Stats
-extern int32 ENGINE_API GCurrentRunningSounds;
+extern int32 AKAUDIO_API GCurrentRunningSounds;
 //</CSS>
 
 /*------------------------------------------------------------------------------------
@@ -4584,3 +4584,7 @@ void FAkAudioDevice::RegisterBusMeteringDelegate(const FString &BusName, EAkMete
 }
 
 // </CSS>
+
+// <MODDING EDIT> These were not defined anywhere
+int GCurrentRunningSounds = 0.0;
+// </MODDING EDIT>
diff --git a/Plugins/Wwise/Source/AkAudio/Private/AkComponent.cpp b/Plugins/Wwise/Source/AkAudio/Private/AkComponent.cpp
index 63e6089e0f..4a7b427b76 100644
--- a/Plugins/Wwise/Source/AkAudio/Private/AkComponent.cpp
+++ b/Plugins/Wwise/Source/AkAudio/Private/AkComponent.cpp
@@ -53,7 +53,7 @@ Copyright (c) 2024 Audiokinetic Inc.
 #endif
 
 //<CSS> Stats
-extern int32 ENGINE_API GCurrentRunningSounds;
+extern int32 AKAUDIO_API GCurrentRunningSounds;
 //</CSS>
 
 /*------------------------------------------------------------------------------------
